### PR TITLE
[AIX] Turn on `#pragma mc_func` check by default

### PR DESCRIPTION
**Pull request by qiongsiwu**
https://github.com/llvm/llvm-project/pull/99888 added a check (and corresponding options) to flag uses of `#pragma mc_func` on AIX. 

This PR turns on the check by default.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/AIX.cpp
Comment (chenzheng1030) [inline clang/lib/Driver/ToolChains/AIX.cpp:565]: Since the preprocessor has a default value, can we avoid the cc1 option if the value here matches the default value inside the preprocessor, like:
```
if (Args.hasArg(options::OPT_fno_err_pragma_mc_func_aix))
   CC1Args.push_back("-fno-err-pragma-mc-func-aix");
```

(no need to pass `-ferr-pragma-mc-func-aix` as it is the default inside preprocessor.)

Comment (qiongsiwu) [reply clang/lib/Driver/ToolChains/AIX.cpp]: Good point! Code is revised. 

I tried the suggested change
```
if (Args.hasArg(options::OPT_fno_err_pragma_mc_func_aix))
   CC1Args.push_back("-fno-err-pragma-mc-func-aix");
```
But it does not handle situations like `-fno-err-pragma-mc-func-aix -ferr-pragma-mc-func-aix` correctly. I think we still need to process both flags in the driver, but we only need to pass `-fno-err-pragma-mc-func-aix` to CC1. Did I miss something?

Comment (chenzheng1030) [reply clang/lib/Driver/ToolChains/AIX.cpp]: Your code is the correct one. Thanks for pointing this out.

**Review by chenzheng1030 — COMMENTED** [review]
LGTM. Please wait for @AaronBallman 's approval. Thanks

**Review by AaronBallman — APPROVED** [review]
LGTM! Please make sure 1) this gets backported to the 19.x branch (https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches), 2) the 19.x release branch gets a release note for the pragma and command line options.

**Comment by qiongsiwu** [comment]
@AaronBallman sorry to bug you. This patch was not tested sufficiently on my end and it is causing build failures on AIX buildbots due to the aforementioned system headers including `#pragma mc_func`. See https://lab.llvm.org/buildbot/#/builders/64/builds/582/steps/5/logs/stdio for example. The compiler-rt cmake files need some updates to add the `no-err` flag. Additionally, there are tests that needs to be updated (we need to augment the tests with `-fno-err-pragma-mc-func-aix` e.g. https://lab.llvm.org/buildbot/#/builders/64/builds/550). Do you think it is reasonable to revert this PR for now (and from llvm19)? We can reland the PR once these necessary changes are made in one batch together? 

Thanks for your input!

**Comment by qiongsiwu** [comment]
Pushed dd7a4c3e5ee3300588b7c12631f3305553d8ea6c to revert this PR for the moment since it is causing too many failures (upstream and downstream). Will revert this PR from llvm19 release branch as well. 

/cherry-pick dd7a4c3e5ee3300588b7c12631f3305553d8ea6c

**Comment by MaskRay** [comment]
(Tips: Personally when I have some patches intended for the release branch, I'll build and test llvm-project in the release branch as well to decrease the risk of potential immediate reverts/churn.)

**Comment by AaronBallman** [comment]
> @AaronBallman sorry to bug you. This patch was not tested sufficiently on my end and it is causing build failures on AIX buildbots due to the aforementioned system headers including `#pragma mc_func`. See https://lab.llvm.org/buildbot/#/builders/64/builds/582/steps/5/logs/stdio for example. The compiler-rt cmake files need some updates to add the `no-err` flag. Additionally, there are tests that needs to be updated (we need to augment the tests with `-fno-err-pragma-mc-func-aix` e.g. https://lab.llvm.org/buildbot/#/builders/64/builds/550). Do you think it is reasonable to revert this PR for now (and from llvm19)? We can reland the PR once these necessary changes are made in one batch together?
> 
> Thanks for your input!

I think it would make sense to revert this and the original patch (so there's no `#pragma mc_func` support in Clang 19.x or main at all) so we can take a fresh run at a new PR to see all of the impacts and consider the design. My concern with reverting just this bit and not the original PR is that we'll leave the original functionality as-is permanently despite the design concerns with it being off-by-default functionality. WDYT?

**Comment by cor3ntin** [comment]
Should we consider a feature like `-Wunknown-pragmas=<pragma-name>` rather than doing something specific for that one pragma?
Have you consider whether this use case generalizes at all?

**Comment by qiongsiwu** [comment]
Sorry about the delay! I was on break. Thanks for the comments @AaronBallman and @cor3ntin ! 

> I think it would make sense to revert this and the original patch (so there's no #pragma mc_func support in Clang 19.x or main at all) so we can take a fresh run at a new PR to see all of the impacts and consider the design.

I have relayed the comments internally and we are discussing it. I will let you know as soon as possible. 

> Should we consider a feature like -Wunknown-pragmas=<pragma-name> rather than doing something specific for that one pragma?
> Have you consider whether this use case generalizes at all?

No we were not aware of other use cases. But `-Wunknown-pragmas=<pragma-name>` sounds like a good idea! Do you think this can be useful in other cases? How do we envision this warning/error to work with the existing `unknown-pragmas` warning if they are both present? I will draft a discourse RFC if this is sufficiently general/useful.

**Comment by AaronBallman** [comment]
> I have relayed the comments internally and we are discussing it. I will let you know as soon as possible.

Thank you!

> No we were not aware of other use cases. But -Wunknown-pragmas=<pragma-name> sounds like a good idea! Do you think this can be useful in other cases? How do we envision this warning/error to work with the existing unknown-pragmas warning if they are both present? I will draft a discourse RFC if this is sufficiently general/useful.

I think this would be useful for both pragmas and attributes (with `-Wunknown-attribute=` as well), but it might be somewhat more involved to implement because IIRC we have to treat those warnings as command line options directly (in Options.td) in order to get the correct behavior.

I would imagine that we'd want last-flag-wins behavior for cases like `-Wunknown-pragmas=foo -Wno-unknown-pragmas` (silences all pragma warnings) or `-Wno-unknown-pragmas -Wunknown-pragmas=foo` (silences all pragma warnings except for `#pragma foo`). Similar for attributes. We probably also want to support a delimited list of pragmas/attributes to ignore, like `-Wno-unknown-attributes=foo,bar,baz`. I don't think we need to care about the syntax used for the attribute (`__attribute__` vs `[[]]` vs `__declspec`), just the attribute name. But it does mean we have to think about cases where the same attribute name is under different prefixes, like `clang::builtin_alias` and `clang_builtin_alias` or `clang::no_stack_protector` and `gnu::no_stack_protector`.

**Comment by AaronBallman** [comment]
> Thank you! However, we're getting pretty late in the release cycle and so I'd like to see this backed out of 19.x ASAP (and given the direction the discussion is going, I think it should be backed out of main as well).

@qiongsiwu will you be able to get these changes reverted today? If not, I'll try to revert them this afternoon. CC @tstellar @tru for release manager awareness.

**Comment by qiongsiwu** [comment]
> > Thank you! However, we're getting pretty late in the release cycle and so I'd like to see this backed out of 19.x ASAP (and given the direction the discussion is going, I think it should be backed out of main as well).
> 
> @qiongsiwu will you be able to get these changes reverted today? If not, I'll try to revert them this afternoon. CC @tstellar @tru for release manager awareness.

Yes I will get to this later today. We have discussed and it is OK at the moment to pull out this check. I will open a PR soon (to revert all the commits relevant to the check).

**Comment by qiongsiwu** [comment]
https://github.com/llvm/llvm-project/pull/102919 is created to revert this feature now. I will review our discussion and consider an improved implementation. Thanks again for the input and discussion!

**Pull request merged**
